Allow calling extension methods on `this` as receiver

dembanakh opened the pull request:
Addresses #3334 

Currently in Javac, whenever the call receiver is `this`, it is not considered as a candidate for extension method call.
https://github.com/projectlombok/lombok/blob/8bb4dc5daa8792e718e88c596812f232a2b73e6e/src/core/lombok/javac/handlers/HandleExtensionMethod.java#L174

This PR makes the behaviour consistent with that of Eclipse's handler, I believe. In other words, explicit `this` call would be considered as a possible extension method call.

dembanakh committed:
Allow calling extension methods on `this` as receiver

dembanakh commented:
@Rawi01 Could you take a look at this by any chance? 😃

Rawi01 commented on test/transform/resource/after-delombok/ExtensionMethodThis.java:
This test fails because the method name is different

Rawi01 reviewed (changes requested):
- Please add your name to the AUTHORS file
- It would be great if you can rebase this branch on the current master. We recently enabled GitHub Actions to run the whole test suite and that way we do not have to run the tests locally. (Some eclipse test will fail but you can ignore that)

I'm not sure if extension methods should ever work like that but it should at least be consistent. It feels a little bit weird because you could simply add the method to your class instead of using extension method at all.

dembanakh commented:
Thanks, I fixed the failing test and rebased.

4 test suites are failing (the same 4 that fail on master), but the others pass successfully, so I hope that's alright.

While it indeed feels strange to call an extension method on `this`, I think this option is worth keeping. Consider e.g. the case where you have to use the extension method both inside and outside the receiver's class
```java
@ExtensionMethod(Extensions.class)
class A {
  public void reset() {
    this.clear();
    // ...
  }
}

@ExtensionMethod(Extensions.class) 
class B {
  private A a;

  public void run() {
    a.clear();
    // ...
  }
}

class Extensions {
  public static void clear(A a) {}
}
```

Rawi01 commented:
In your example you could also add the `clear` method to `A` but this still might be useful if the base class is a third party/framework class.
